File: p019m/remove_nth_from_end.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from local_packages.list import ListNode


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0, head)
        pf, pb = dummy, dummy
        while n >= 0:
            pf, n = pf.next, n - 1
        while pf:
            pf, pb = pf.next, pb.next
        pb.next = pb.next.next
        return dummy.next


# TESTS
for array, n, expected in [
    ([1, 2, 3, 4, 5], 2, [1, 2, 3, 5]),
    ([1], 1, []),
    ([1, 2], 1, [1]),
]:
    sol = Solution()
    actual = ListNode.to_array(sol.removeNthFromEnd(ListNode.from_array(array), n))
    print(f"Remove {n}th node from end of {array} -> {actual}")
    assert actual == expected

File: p019m/remove_nth_from_end.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
const List = require('list');
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  // Create the dummy head to deal with the case of remove the only node in the list
  let dummy = new List.ListNode(0);
  dummy.next = head;
  let [fst, snd] = [dummy, dummy];
  for (let i = 1; i <= n + 1; ++i) fst = fst.next;
  while (fst) {
    [fst, snd] = [fst.next, snd.next];
  }
  snd.next = snd.next.next;
  return dummy.next;
};
// TESTS
[
  [[5], 1, []],
  [[1, 2], 1, [1]],
  [[1, 2], 2, [2]],
  [[1, 2, 3, 4, 5], 1, [1, 2, 3, 4]],
  [[1, 2, 3, 4, 5], 2, [1, 2, 3, 5]],
  [[1, 2, 3, 4, 5], 5, [2, 3, 4, 5]],
].forEach(([array, n, expected]) => {
  let actual = List.toArray(removeNthFromEnd(List.fromArray(array), n));
  console.log(`Remove ${n}th node from end of ${array} -> ${actual}`);
  console.assert(actual.toString() === expected.toString());
});
